compiler/rustc_trait_selection/src/traits/project.rs

```diff
@@ -1294,6 +1294,23 @@ fn assemble_candidate_for_impl_trait_in_trait<'cx, 'tcx>(
         let _ = selcx.infcx.commit_if_ok(|_| {
             match selcx.select(&obligation.with(tcx, trait_predicate)) {
                 Ok(Some(super::ImplSource::UserDefined(data))) => {
+                    let Ok(leaf_def) = specialization_graph::assoc_def(tcx, data.impl_def_id, trait_fn_def_id) else {
+                        return Err(());
+                    };
+                    // Only reveal a specializable default if we're past type-checking
+                    // and the obligation is monomorphic, otherwise passes such as
+                    // transmute checking and polymorphic MIR optimizations could
+                    // get a result which isn't correct for all monomorphizations.
+                    if !leaf_def.is_final()
+                        && obligation.param_env.reveal() != Reveal::All
+                        && selcx
+                            .infcx
+                            .resolve_vars_if_possible(obligation.predicate.trait_ref(tcx))
+                            .still_further_specializable()
```

Comment:
that condition seems wrong :thinking: it should be `!leaf_def.is_final() && (obligation.param_env.reveal() != Reveal::All || selcx.infcx.resolve_vars_if_possible(obligation.predicate.trait_ref(tcx))
                            .still_further_specializable())`

or probably nicer to state this positively, we only return `Ok` if

```rust
leaf_def.is_final()
|| (
    obligation.param_env.reveal() == Reveal::All
    && !selcx.infcx.resolve_vars_if_possible(
        obligation.predicate.trait_ref(tcx)
    ).still_further_specializable()
)
```